kernel/ps2/src/lib.rs

```diff
@@ -1,484 +1,704 @@
+/// otherwise would read bogus data
+fn read_controller_test_result() -> Result<ControllerTestResult, &'static str> {
+    read_data().try_into().map_err(|_| "failed to read controller test result")
+}
 
-        // set the new configuration
-        let new_config = (old_config & 0xDF) | 0x02;
-        ps2_write_config(new_config);
-    }
-    {
-        ps2_write_command(0xAE);
-        ps2_write_command(0xA8);
-    }
+#[derive(TryFromPrimitive)]
+#[repr(u8)]
+enum PortTestResult {
+    Passed = 0x00,
+    ClockLineStuckLow = 0x01,
+    ClockLineStuckHigh = 0x02,
+    DataLineStuckLow = 0x03,
+    DataLineStuckHigh = 0x04,
 }
 
-/// test the second ps2 data port
-pub fn test_ps2_port2() {
-    // test the ps2 controller
-    {
-        ps2_write_command(0xAA);
-        let test_flag = ps2_read_data();
-        if test_flag == 0xFC {
-            warn!("ps2 controller test failed!!!")
-        } else if test_flag == 0x55 {
-            info!("ps2 controller test pass!!!")
-        }
-    }
-    // test the second ps2 data port
-    {
-        ps2_write_command(0xA9);
-        let test_flag = ps2_read_data();
-        match test_flag {
-            0x00 => info!("second ps2 port test pass!!!"),
-            0x01 => warn!("second ps2 port clock line stuck low"),
-            0x02 => warn!("second ps2 port clock line stuck high"),
-            0x03 => warn!("second ps2 port data line stuck low"),
-            0x04 => warn!("second ps2 port data line stuck high"),
-            _ => {}
-        }
-    }
-    // enable PS2 interrupt and see the new config
-    {
-        let config = ps2_read_config();
-        if config & 0x02 == 0x02 {
-            info!("second ps2 port interrupt enabled")
-        } else {
-            warn!("second ps2 port interrupt disabled")
-        }
-        if config & 0x20 != 0x20 {
-            info!("second ps2 data port enabled")
-        } else {
-            warn!("second ps2 data port disabled")
+/// must only be called after writing the [TestPort1] or [TestPort2] command
+/// otherwise would read bogus data
+fn read_port_test_result() -> Result<PortTestResult, &'static str> {
+    read_data().try_into().map_err(|_| "failed to read port test result")
+}
+
+// https://wiki.osdev.org/PS/2_Keyboard#Special_Bytes //NOTE: all other bytes sent by the keyboard are scan codes
+// http://users.utcluj.ro/~baruch/sie/labor/PS2/PS-2_Mouse_Interface.htm mouse only returns AA, FC or FA, mouse_id and packet
+#[derive(TryFromPrimitive)]
+#[repr(u8)]
+pub enum DeviceToHostResponse {
+    KeyDetectionErrorOrInternalBufferOverrun1 = 0x00,
+    SelfTestPassed = 0xAA, //sent after "0xFF (reset)" command or keyboard power up
+    ResponseToEcho = 0xEE,
+    Acknowledge = 0xFA,
+    SelfTestFailed1 = 0xFC, //sent after "0xFF (reset)" command or keyboard power up
+    SelfTestFailed2 = 0xFD, //sent after "0xFF (reset)" command or keyboard power up
+    ResendCommand = 0xFE,
+    KeyDetectionErrorOrInternalBufferOverrun2 = 0xFF,
+}
```

Comment:
all of these comments would be better off as doc comments such that rustdoc can display them, otherwise they're invisible to developers and folks browsing the docs. Same with the other struct/enum definitions herein.